rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp

```diff
@@ -32,6 +32,12 @@ namespace rclcpp
 namespace executors
 {
 
+/**
+ * The Multi-threaded executor uses multiple threads to do work while spinning.
+ *
+ * Note, however, that it will only use multiple threads via its blocking `spin()` method.
+ * When using other methods, like e.g. `rclcpp::Executor:spin_until_future_complete(some_future)`, only the main thread will be used to do work.
+ */
```

Comment:
how about adding the following on top of `MultiThreadedExecutor::spin`?

```cpp
  /// Multiple-threaded implementation of spin.
  /**
   * This function will block until work comes in, and try to execute all work with
   * configured multiple threads, and then repeat the process until canceled.
   * It may be interrupt by a call to rclcpp::Executor::cancel() or by ctrl-c
   * if the associated context is configured to shutdown on SIGINT.
   * \throws std::runtime_error when spin() called while already spinning
   */
```

i think `rclcpp::Executor:spin_until_future_complete(some_future)` explains it good enough.

Reply:
Thanks for reviewing! I am happy to add your suggestion to the spin method as part of this PR. :+1:

However, I don't think it addresses the point of my PR. I don't want to push this too hard. It's fine if you think it's unnecessary / out of place. I just want to clarify what exactly my point was:
While it explains that the `spin()` method will use multiple threads, it does not clarify that **only**(?) the `spin()` method will work with multiple threads. My (naive) expectation was that **all** spin-related methods I call on the `MultiThreadedExecutor` would run with multiple threads. Including methods inherited from the `rclcpp::Executor` (maybe except some, like `spin_once`, which explicitly only work on a single item).
Did I miss something that could have changed my expectation?

The documentation of [`rclcpp::Executor:spin_until_future_complete(some_future)`](https://docs.ros2.org/latest/api/rclcpp/classrclcpp_1_1Executor.html#a484f01fbbbdc2604f349586d2b2f6830) does not talk about multiple threads ("`Spin (blocking) until the future is complete, it times out waiting, or rclcpp is interrupted."`). I think that's fine, since it is probably nice if the base class does not need to know about its specializations' properties. So adding it there felt out of place, which is why I added the comment to the `MultiThreadedExecutor` class.

In the end, I ended up inspecting the implementation of the `MultiThreadedExecutor` and its base class `rclcpp::Executor` to see that the multi threading stuff is implemented only in the overridden `spin()` method and the methods the `Executor` implements do **not** use `virtual void spin()` (and are not overridden by the `MultiThreadedExecutor`).

I have found that another method, [spin_some](https://docs.ros2.org/latest/api/rclcpp/classrclcpp_1_1Executor.html#ae821cd4cf1040fe2da521fb19a458fa1), actually has a note about the default implementation only being suitable for a single-threaded model of execution. Which was what tipped me off while looking for the source of my unexpected behavior ("maybe other methods of the `rclcpp::Executor` are also only suitable for single-threaded use. Does the `MultiThreadedExecutor` have a specialized implementation? (no)").

Reply:
thanks for the explanation.

i see what you mean here. i was thinking that under the class `MultiThreadedExecutor` only implements `spin` so it would be obvious for the user. but that is the problem here.

> maybe other methods of the rclcpp::Executor are also only suitable for single-threaded use

probably we would want to add more documentation for `rclcpp::Executor` here about single-threaded? because that is the behavior for `rclcpp::Executor`, and then user can see that is inherited from `rclcpp::Executor`.

i am not inclined to add docs like `only the main thread will be used to do work` to `MultiThreadedExecutor`. if we have `MultiThreadedXXXExecutor`, we need to add that every time. instead, i think we can add the docs where that behavior is actually implemented?

Reply:
TL;DR: I will add a note to `rclcpp::Executor`'s `spin_until_future_complete` method. This would have saved past-me from the mistakes I made. :)

> i am not inclined to add docs like only the main thread will be used to do work to MultiThreadedExecutor. if we have MultiThreadedXXXExecutor, we need to add that every time. instead, i think we can add the docs where that behavior is actually implemented?
True, good point, that would be annoying.

> class MultiThreadedExecutor only implements spin so it would be obvious for the user. but that is the problem here.
You're right, I was jumping to conclusions. I just saw a method that sounded like it would do what I wanted and I tried to use it without much consideration.
But by just looking at the headers/docs, I do not think it is obvious that `spin_until_future_complete` does not use a virtual method that might be implemented by the `MultiThreadedExecutor` such that you would get multi-threaded execution.
Of course, after seeing that only `spin()` is overwritten, and considering how spin() behaves, it should be fairly obvious.